In [1]:
import pandas as pd
df = pd.read_csv("../data/raw/insurance.csv")
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [2]:
df.shape

(1338, 7)

In [3]:
df.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='str')

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 94.5 KB


In [5]:
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(1)

In [7]:
df.describe()

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010


In [8]:
df["sex"].value_counts()

sex
male      676
female    662
Name: count, dtype: int64

In [9]:
df["smoker"].value_counts()

smoker
no     1064
yes     274
Name: count, dtype: int64

In [10]:
df["region"].value_counts()

region
southeast    364
southwest    325
northwest    325
northeast    324
Name: count, dtype: int64

In [11]:
print("Sex:", df["sex"].unique())
print("Smoker:", df["smoker"].unique())
print("Region:", df["region"].unique())

Sex: <ArrowStringArray>
['female', 'male']
Length: 2, dtype: str
Smoker: <ArrowStringArray>
['yes', 'no']
Length: 2, dtype: str
Region: <ArrowStringArray>
['southwest', 'southeast', 'northwest', 'northeast']
Length: 4, dtype: str


In [12]:
df.groupby("smoker")["charges"].mean()

smoker
no      8434.268298
yes    32050.231832
Name: charges, dtype: float64

In [13]:
df.groupby("age")["charges"].mean().head(10)

age
18     7086.217556
19     9747.909335
20    10159.697736
21     4730.464330
22    10012.932802
23    12419.820040
24    10648.015962
25     9838.365311
26     6133.825309
27    12184.701721
Name: charges, dtype: float64

In [14]:
df[["bmi", "charges"]].corr()

,bmi,charges
bmi,1.000000,0.198341
charges,0.198341,1.000000


In [15]:
df[["age", "charges"]].corr()

,age,charges
age,1.000000,0.299008
charges,0.299008,1.000000


In [16]:
df.groupby("children")["charges"].mean()

children
0    12365.975602
1    12731.171832
2    15073.563734
3    15355.318367
4    13850.656311
5     8786.035247
Name: charges, dtype: float64

In [17]:
import plotly.express as px
fig = px.histogram(
   df,
   x="charges",
   nbins=40,
   title="Distribution of Medical Charges",
   labels={"charges": "Medical Charges"}
)
fig.show()

In [18]:
df.groupby("smoker")["charges"].mean()

smoker
no      8434.268298
yes    32050.231832
Name: charges, dtype: float64

In [19]:
fig = px.box(
   df,
   x="smoker",
   y="charges",
   color="smoker",
   title="Medical Charges: Smoker vs Non-Smoker",
   labels={
       "smoker": "Smoker",
       "charges": "Medical Charges"
   }
)
fig.show()

In [20]:
df.groupby("age")["charges"].mean().head(10)

age
18     7086.217556
19     9747.909335
20    10159.697736
21     4730.464330
22    10012.932802
23    12419.820040
24    10648.015962
25     9838.365311
26     6133.825309
27    12184.701721
Name: charges, dtype: float64

In [21]:
df.groupby("age")["charges"].mean().head(10) 

age
18     7086.217556
19     9747.909335
20    10159.697736
21     4730.464330
22    10012.932802
23    12419.820040
24    10648.015962
25     9838.365311
26     6133.825309
27    12184.701721
Name: charges, dtype: float64

In [22]:
fig = px.scatter(
   df,
   x="age",
   y="charges",
   color="smoker",
   title="Age vs Medical Charges",
   labels={
       "age": "Age",
       "charges": "Medical Charges"
   }
)
fig.show()

In [23]:
df[["bmi", "charges"]].corr()

,bmi,charges
bmi,1.000000,0.198341
charges,0.198341,1.000000


In [24]:
fig = px.scatter(
   df,
   x="bmi",
   y="charges",
   color="smoker",
   title="BMI vs Medical Charges",
   labels={
       "bmi": "BMI",
       "charges": "Medical Charges"
   }
)
fig.show()

In [25]:
df.groupby("children")["charges"].mean()

children
0    12365.975602
1    12731.171832
2    15073.563734
3    15355.318367
4    13850.656311
5     8786.035247
Name: charges, dtype: float64

In [26]:
fig = px.box(
   df,
   x="children",
   y="charges",
   title="Medical Charges by Number of Children",
   labels={
       "children": "Number of Children",
       "charges": "Medical Charges"
   }
)
fig.show()

In [27]:
df.groupby("region")["charges"].mean().sort_values(ascending=False)

region
southeast    14735.411438
northeast    13406.384516
northwest    12417.575374
southwest    12346.937377
Name: charges, dtype: float64

In [28]:
fig = px.box(
   df,
   x="region",
   y="charges",
   title="Medical Charges by Region",
   labels={
       "region": "Region",
       "charges": "Medical Charges"
   }
)
fig.show()

In [29]:
df[["age", "bmi", "children", "charges"]].corr()

,age,bmi,children,charges
age,1.000000,0.109272,0.042469,0.299008
bmi,0.109272,1.000000,0.012759,0.198341
children,0.042469,0.012759,1.000000,0.067998
charges,0.299008,0.198341,0.067998,1.000000


In [30]:
high_cost = df[df["charges"] > df["charges"].quantile(0.75)]
high_cost.shape

(335, 7)

In [31]:
high_cost.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
3,33,male,22.705,0,no,northwest,21984.47061
9,60,female,25.840,0,no,northwest,28923.13692
11,62,female,26.290,0,yes,southeast,27808.72510
14,27,male,42.130,0,yes,southeast,39611.75770


## EDA Findings

- Smoking status shows a strong difference in medical charges.
- Age has a relationship with medical cost.
- BMI shows some relationship with medical charges.
- Medical charges have a highly right-skewed distribution.
- Some customers have significantly higher medical costs.

In [32]:
df_clean = df.copy()
df_clean = df_clean.drop_duplicates()
df_clean.shape

(1337, 7)

In [33]:
df_clean.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [34]:
print(df_clean["age"].min())
print(df_clean["age"].max())
print(df_clean["bmi"].min())
print(df_clean["bmi"].max())
print(df_clean["charges"].min())
print(df_clean["charges"].max())

18
64
15.96
53.13
1121.8739
63770.42801


In [35]:
df_clean.to_csv(
   "../data/processed/insurance_clean.csv",
   index=False
)

In [36]:
clean_df = pd.read_csv("../data/processed/insurance_clean.csv")
print("Shape:", clean_df.shape)
print("\nMissing Values:")
print(clean_df.isnull().sum())
print("\nDuplicate Rows:", clean_df.duplicated().sum())

Shape: (1337, 7)

Missing Values:
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

Duplicate Rows: 0


In [37]:
clean_df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [38]:
clean_df.describe()

,age,bmi,children,charges
count,1337.000000,1337.000000,1337.000000,1337.000000
mean,39.222139,30.663452,1.095737,13279.121487
std,14.044333,6.100468,1.205571,12110.359656
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.290000,0.000000,4746.344000
50%,39.000000,30.400000,1.000000,9386.161300
75%,51.000000,34.700000,2.000000,16657.717450
max,64.000000,53.130000,5.000000,63770.428010


## Data Cleaning Summary

- Original records: 1338
- Duplicate records removed: 1
- Final records: 1337
- Missing values: None
- Raw dataset preserved separately
- Clean dataset saved as insurance_clean.csv